sql: add constraint IDs and use them for comments

fqazi opened the pull request:
This pull request will make the following changes:

1) Modify table descriptor logic to start allocating constraint IDs for each table constraint.
2) Add a migration to upgrade existing descriptors to have constraint IDs
3) Modify the comment updater / comment logic to references constraints by ConstraintID and update the obj_description builtin to look these up.

fqazi committed:
catalog: add the ConstraintID for table descriptors

Previously, we had no good way of associating
back to the table descriptor for both the declarative
schema changer and comments on constraint. This was
inadequate because we would use hashing schemes to associate
comments (OID's) back to constraints. Additionally, since no
indexing scheme exists, renames in declarative schema changer
would be impossible to implement. To address this, this patch
starts off by adding ConstraintIDs for new table descriptors
and during post deserialization of descriptors.
ConstraintID will be an increasing number assigned to each
constraint in a table descriptor.

Release note: None

fqazi committed:
migration: add migration for constraint ID and version gate

Previously, after adding constraint ID there was no version
gate logic to ensure that this new field would be propagated
to all table descriptors, making it dangerous to comment on
constraint when the cluster was not fully upgraded
(i.e. the IDs could shift as descriptors were modified).
To address this patch introduces a new migration to upgrade
descriptors to have the constraint ID field and additionally,
block comments on constraints until the new version is active
on the cluster.

Release note: None

fqazi committed:
sql: remove OID logic from comment updater

Previously, the comment updater had logic for computing
OIDs for constraints. This logic is now unnecessary since
constraints are stored by ConstraintID now. To address this,
this patch removes the OID logic from the comment updater.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75499)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
quick first pass on first commit 

> Reviewed 7 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-:-MuGxtmQ-sYOeUyu3S3G:bfkcm0v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @postamar, @rafiss, and @RichardJCai)

___
*[pkg/sql/catalog/descpb/structured.proto, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MuGwyJe16P6s5LepfEw:-MuGwyJf0M-WBIkSoGrj:b-2glzfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/0452e0300a238233e6fcc24d727d61789ef4f4e3/pkg/sql/catalog/descpb/structured.proto#L97)):*
> ```protobuf
>     (gogoproto.casttype) = "ConstraintID", (gogoproto.nullable) = false];
>   // These fields were used for foreign keys until 20.1.
>   reserved 11, 12, 13;
> ```

I don't think you can re-use a reserved field number. Instead make a new one. 

Can I request that you comment all of these constraint_id fields? 



_Code quote:_
```protobuf
  optional uint32 constraint_id = 10 [(gogoproto.customname) = "ConstraintID",
    (gogoproto.casttype) = "ConstraintID", (gogoproto.nullable) = false];
  // These fields were used for foreign keys until 20.1.
  reserved 11, 12, 13;
```
___
*[pkg/sql/catalog/descpb/structured.proto, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MuGxEH60JPHAGX5Myy8:-MuGxEH60JPHAGX5Myy9:bpt385r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0452e0300a238233e6fcc24d727d61789ef4f4e3/pkg/sql/catalog/descpb/structured.proto#L521)):*
> ```protobuf
>   optional bool use_delete_preserving_encoding = 24 [(gogoproto.nullable) = false];
> 
>   optional uint32 constraint_id = 33 [(gogoproto.customname) = "ConstraintID",
> ```

This one is particularly subtle and worth commenting. It might be a unique constraint, it might be a primary key constraint. Not all indexes will be constraints. See the `created_explicitly` field (which is deeply buggy :/). 
___
*[pkg/sql/catalog/tabledesc/table_desc_builder.go, line 616 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MuGxZrC9S_MYqZVu3ZD:-MuGxZrC9S_MYqZVu3ZE:b-xtfw1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0452e0300a238233e6fcc24d727d61789ef4f4e3/pkg/sql/catalog/tabledesc/table_desc_builder.go#L616)):*
> ```Go
> 		desc.NextConstraintID = 1
> 	}
> 	// Loop over all constraints and assign constraint IDs.
> ```

I think we need to deal with the mutations too. Especially interesting is when we have `ALTER PRIMARY KEY`. I think the ID should be the same on both. That might be tricky. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75499)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @postamar, @rafiss, and @RichardJCai)

___
*[pkg/sql/catalog/descpb/structured.proto, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MuGwyJe16P6s5LepfEw:-MuHCV6W1LG268AtlxH7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0452e0300a238233e6fcc24d727d61789ef4f4e3/pkg/sql/catalog/descpb/structured.proto#L97)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think you can re-use a reserved field number. Instead make a new one. 

Can I request that you comment all of these constraint_id fields? 
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/table_desc_builder.go, line 616 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MuGxZrC9S_MYqZVu3ZD:-MuHCX849uDvTU5AAJQs:b-cm6210) ([raw file](https://github.com/cockroachdb/cockroach/blob/0452e0300a238233e6fcc24d727d61789ef4f4e3/pkg/sql/catalog/tabledesc/table_desc_builder.go#L616)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think we need to deal with the mutations too. Especially interesting is when we have `ALTER PRIMARY KEY`. I think the ID should be the same on both. That might be tricky. 
</blockquote></details>

Done.

I modified the ALTER PRIMARY KEY code and migration logic to handle this by moving the values across matching index IDs.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Haven't gotten through it but here's an immediate observation. 

> Reviewed 1 of 16 files at r1, 1 of 27 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-:-MuRVtzP-pGvm0fn9R16:b3e0q3o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @postamar, @rafiss, and @RichardJCai)

___
*[pkg/sql/catalog/systemschema/system.go, line 964 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MuNAt50DUnsoCRzPr60:-MuNAt50DUnsoCRzPr61:bnl1wyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a94ac074ca3db34743074984b48db8d4c27099f2/pkg/sql/catalog/systemschema/system.go#L964)):*
> ```Go
> 			tbl.NextMutationID = 0
> 			tbl.NextConstraintID = 0
> 			tbl.PrimaryIndex.ConstraintID = 0
> ```

should these be 1?



_Code quote:_
```Go
			tbl.NextConstraintID = 0
			tbl.PrimaryIndex.ConstraintID = 0
```
___
*[pkg/cli/testdata/doctor/test_examine_zipdir_verbose, line 42 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MuRUpY9AsWyf-fVnOXy:-MuRUpY9AsWyf-fVnOXz:bv41xyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/11da0cfde9f998edd871a497a54fee457cf4a943/pkg/cli/testdata/doctor/test_examine_zipdir_verbose#L42)):*
> ```
>   ParentID  52, ParentSchemaID 29: relation "users" (53): referenced database ID 52: referenced descriptor not found
>   ParentID  52, ParentSchemaID 29: relation "users" (53): processed
>   ParentID  52, ParentSchemaID 29: relation "vehicles" (54): referenced database ID 52: referenced descriptor not found
> ```

these are suspect, I think we'll want to do more to gate the validation of this migration. A cluster version plumbed into validation is probably part of it. I suspect it'll take a bit of thought on how we want to design this. 

One straw man idea would be to have the version of the cluster plumbed into validate dictate whether constraint ids must be on, and then, if they are on, validate as you are here. We want to plumb a current version into validation anyway, it's something @postamar brought up recently. This is a good motivation. It then begs the question of what to do for doctor. The right thing would be to read it out of `system.settings`, but that's a bit involved. In the short term, we could hard-code it for these tests and have a flag. I'd be happy with that. 


<!-- Sent from Reviewable.io -->

postamar reviewed (changes requested):
I did a review pass. The constraint ID changes are high-quality and mostly good, but I think your version gating change isn't quite there yet, and feel out of scope of this PR despite it being dependent on them. 


I suggest you remove that last commit from this PR, rework it and submit it for review for it to be merged ahead of this PR.

> Reviewed 5 of 16 files at r1, 1 of 3 files at r11, 2 of 14 files at r12, 1 of 3 files at r13, 2 of 11 files at r14, 1 of 36 files at r18, 1 of 37 files at r19, 174 of 174 files at r24, 24 of 24 files at r25, 9 of 9 files at r26, 12 of 12 files at r27, 135 of 135 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-:-Mv0JJVzDojZnC_4bgLX:bhwor2s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @fqazi, @rafiss, and @RichardJCai)

___
*[pkg/cli/doctor.go, line 94 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0EP8S6k6ZeBEIL-dR:-Mv0EP8S6k6ZeBEIL-dS:b-83v4r2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/cli/doctor.go#L94)):*
> ```Go
> requires the path of the unzipped 'debug' directory as its argument. A version
> can be optionally specified, which will be used enable / disable validation
> that may not exist on downlevel versions.
> ```

Neat!
___
*[pkg/cli/doctor.go, line 169 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0ESiS4OTcv9SMlLNc:-Mv0ESiS4OTcv9SMlLNd:b-3ykv4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/cli/doctor.go#L169)):*
> ```Go
> 	if version == nil {
> 		version = &clusterversion.ClusterVersion{
> 			Version: clusterversion.ByKey(clusterversion.ConstraintIDsForTableDescs),
> ```

Shouldn't the default cluster version be the last one instead?
___
*[pkg/migration/migrations/ensure_constraint_id.go, line 24 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0BZxm1XFMvs5wAYHr:-Mv0BZxm1XFMvs5wAYHs:b8yk4hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/69d68af2469f9904c09b6b1fb4fcf0f3146ef8b9/pkg/migration/migrations/ensure_constraint_id.go#L24)):*
> ```Go
> )
> 
> func ensureTablesHasConstraintID(ctx context.Context, _ clusterversion.ClusterVersion, d migration.TenantDeps, _ *jobs.Job) error {
> ```

`ensureTablesHaveConstraintIDs`?
___
*[pkg/migration/migrations/ensure_constraint_id.go, line 63 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0BkQo5_SM1yM6ugTT:-Mv0BkQo5_SM1yM6ugTU:bml1vdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/69d68af2469f9904c09b6b1fb4fcf0f3146ef8b9/pkg/migration/migrations/ensure_constraint_id.go#L63)):*
> ```Go
> 			return txn.Run(ctx, batch)
> 		})
> }
> ```

Is this migration even necessary? I suspect that Richard's privileges migration will already be upgrading all table descriptors making this one entirely redundant. If that's not the case, then we should modify the existing migration to make it do what we want. Rewriting the whole descriptors table is expensive, no need to do it twice.
___
*[pkg/sql/create_sequence.go, line 20 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0Emjw6xIesYWjJ7oy:-Mv0Emjw6xIesYWjJ7oz:b-47f1ub) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/create_sequence.go#L20)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/kv"
> 	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
> 	clustersettings "github.com/cockroachdb/cockroach/pkg/settings/cluster"
> ```

Is this deliberate?
___
*[pkg/sql/create_sequence.go, line 244 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0Ev5z0-75kOPPyi1Y:-Mv0Ev5z0-75kOPPyi1Z:b-3h6mfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/create_sequence.go#L244)):*
> ```Go
> 	ctx context.Context,
> 	p *planner,
> 	settings *clustersettings.Settings,
> ```

This seems redundant, isn't the planner able to provide the cluster settings already?
___
*[pkg/sql/create_table.go, line 1943 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0F47_B_5pI1B-7vH0:-Mv0F47_B_5pI1B-7vH1:bvy1g2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/create_table.go#L1943)):*
> ```Go
> 
> 	if st == nil {
> 		panic("blah")
> ```

detritus?
___
*[pkg/sql/pg_catalog.go, line 1544 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv09w7aBifCgjwJ3gxc:-Mv09w7aBifCgjwJ3gxd:b-57au6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c2ee0704d70e65e00ee168252f1b5a359f98d59/pkg/sql/pg_catalog.go#L1544)):*
> ```Go
> 				classOid = tree.NewDOid(catconstants.PgCatalogClassTableID)
> 			case keys.ConstraintCommentType:
> 				// FIXME: fetch the schema and database
> ```

Is this comment spurious?
___
*[pkg/sql/pg_catalog.go, line 1613 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0A4auAA4GsEaeFe9j:-Mv0A4av3cV-YrOLT1w4:b-92c7ub) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c2ee0704d70e65e00ee168252f1b5a359f98d59/pkg/sql/pg_catalog.go#L1613)):*
> ```Go
> 					}
> 
> 				}
> ```

It might be worth moving this if-else-block to its own helper function which takes `constraint`, the db ID, schema name and table ID, and returns an oid.
___
*[pkg/sql/table.go, line 298 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0FB4T9fCzcXg6-Pdh:-Mv0FB4T9fCzcXg6-Pdi:blqw0tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/table.go#L298)):*
> ```Go
> 
> 	if err := descbuilder.ValidateSelf(tableDesc,
> 		p.ExecCfg().Settings.Version.ActiveVersion(ctx)); err != nil {
> ```

Can you make your whitespacing style at least somewhat consistent please? Having at least some regularity would make for a more pleasant code reading experience.
___
*[pkg/sql/virtual_schema.go, line 251 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0FrvFDA0tfhkkupdm:-Mv0FrvFDA0tfhkkupdn:b-rz7un1) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/virtual_schema.go#L251)):*
> ```Go
> 		Settings: st,
> 	}
> ```

FYI I may be wrong but IIRC there exist some factory methods for similar eval contexts, which we use in the new schema changer dependencies. 

It feels weird to create this eval context only to pass the settings around though. Isn't there a better way?
___
*[pkg/sql/virtual_schema.go, line 697 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0GI3d2oVXMAR3KW7G:-Mv0GI3d2oVXMAR3KW7H:b-fwj1hc) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/virtual_schema.go#L697)):*
> ```Go
> 			td := tabledesc.NewBuilder(&tableDesc).BuildImmutableTable()
> 			if err := descbuilder.ValidateSelf(td,
> 				st.Version.ActiveVersionOrEmpty(ctx)); err != nil {
> ```

This is surprising. Is this right?
___
*[pkg/sql/catalog/descriptor.go, line 516 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv04ux09NRkKjxz7e63:-Mv04ux09NRkKjxz7e64:blw0b5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/descriptor.go#L516)):*
> ```Go
> 	GetNextColumnID() descpb.ColumnID
> 	// GetNextConstraintID returns the next unused constraint ID for this table.
> 	// Constraint ID's are unique per table, but not unique globally.
> ```

nit: remove `'`
___
*[pkg/sql/catalog/table_elements.go, line 130 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv05PkN0TmXu6Ik0TGZ:-Mv05PkO739vDpQRBx8B:b-af0ozv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/table_elements.go#L130)):*
> ```Go
> 
> 	GetID() descpb.IndexID
> 	GetConstraintID() descpb.ConstraintID
> ```

Do we also need a `GetConstraintID()` method on the `ConstraintToUpdate` interface?
___
*[pkg/sql/catalog/descs/descriptor.go, line 131 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0GiX1AfAB6rhLcvM-:-Mv0GiX1AfAB6rhLcvM0:b-wgsrwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/catalog/descs/descriptor.go#L131)):*
> ```Go
> 				ctx,
> 				txn,
> 				tc.settings.Version.ActiveVersion(ctx),
> ```

Shouldn't the settings be a field in the `kv` struct instead?
___
*[pkg/sql/catalog/descs/direct.go, line 145 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0H4MG6kvW2gXyh7m3:-Mv0H4MG6kvW2gXyh7m4:bjf0uig) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/catalog/descs/direct.go#L145)):*
> ```Go
> 	ctx context.Context, txn *kv.Txn, id descpb.ID,
> ) (catalog.DatabaseDescriptor, error) {
> 	desc, err := catkv.MustGetDescriptorByID(ctx, txn, d.codec, d.settings.Version.ActiveVersion(ctx), id, catalog.Database)
> ```

This ActiveVersion retrieval ends up being all over the place, can't we push it down closer to the validation code perhaps? 

Comment applies elsewhere also, of course.
___
*[pkg/sql/catalog/desctestutils/descriptor_test_utils.go, line 31 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0HKup-PBfHzmFNzFz:-Mv0HKup-PBfHzmFNzG-:b72yw2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/catalog/desctestutils/descriptor_test_utils.go#L31)):*
> ```Go
> var LatestClusterVersionForValidationForTest = clusterversion.ClusterVersion{
> 	Version: clusterversion.ByKey(clusterversion.ConstraintIDsForTableDescs),
> }
> ```

Again, we almost certainly have a `var` in `clusterversion` which defines the latest version.
___
*[pkg/sql/catalog/desctestutils/descriptor_test_utils.go, line 70 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0HVU5DAONUszTDcdi:-Mv0HVU5DAONUszTDcdj:bo5cwx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/catalog/desctestutils/descriptor_test_utils.go#L70)):*
> ```Go
> 	kvDB *kv.DB,
> 	codec keys.SQLCodec,
> 	version clusterversion.ClusterVersion,
> ```

Shouldn't these be pointers? Comment applies elsewhere also.
___
*[pkg/sql/catalog/tabledesc/structured.go, line 103 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv06Er92L_56vA0MYnO:-Mv06Er92L_56vA0MYnP:b-kkxzry) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/structured.go#L103)):*
> ```Go
> 	// UpgradedConstraintID indicates that table descriptors had constraint ID
> 	// added.
> 	UpgradedConstraintIDs bool
> ```

Consider naming this `AddedConstraintIDs` in that case?
___
*[pkg/sql/catalog/tabledesc/table_desc_builder.go, line 632 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv06lYY1_bGT5UV9Ted:-Mv06lYY1_bGT5UV9Tee:b4wegew) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/table_desc_builder.go#L632)):*
> ```Go
> 			idx.ConstraintID = desc.GetNextConstraintID()
> 			constraintIndexes[idx.ID] = idx
> 			desc.NextConstraintID++
> ```

Consider defining a closure which will return the ID and increment `desc.NextConstraintID` as a side effect.
___
*[pkg/sql/catalog/tabledesc/validate.go, line 538 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv07uUoDNX6yzV5Rtsu:-Mv07uUoDNX6yzV5Rtsv:b-q9ddwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/validate.go#L538)):*
> ```Go
> 		}
> 	}
> 	return nil
> ```

Consider renaming this `validateConstraintIDs`?
___
*[pkg/sql/descmetadata/metadata_updater.go, line 29 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0DQ830j-DZVeA9x-I:-Mv0DQ830j-DZVeA9x-J:b334w54) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4f36ffed911d6db7a2426a22fc2f21ba821c396/pkg/sql/descmetadata/metadata_updater.go#L29)):*
> ```Go
> )
> 
> // commentUpdater which implements scexec.CommentUpdater that is used to update
> ```

s/commentUpdater/metadataUpdater/ ?
___
*[pkg/sql/schemachanger/scpb/elements.proto, line 406 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0Am-F2jYjdk9vGIOE:-Mv0Am-F2jYjdk9vGIOF:bqlwfm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c2ee0704d70e65e00ee168252f1b5a359f98d59/pkg/sql/schemachanger/scpb/elements.proto#L406)):*
> ```protobuf
>   option (gogoproto.equal) = true;
>   uint32 table_id = 1 [(gogoproto.customname) = "TableID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.ID"];
>   uint32 constraint_id =2 [(gogoproto.customname) = "ConstraintID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.ConstraintID"];
> ```

nit: missing whitespace
___
*[pkg/cli/testdata/doctor/test_examine_zipdir, line 11 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv08pusCcthxOXtR-yU:-Mv08pusCcthxOXtR-yV:b-dhyj2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/cli/testdata/doctor/test_examine_zipdir#L11)):*
> ```
>   ParentID  52, ParentSchemaID 29: relation "vehicle_location_histories" (56): constraint id was missing for constraint: FOREIGN KEY with name fk_city_ref_rides
>   ParentID  52, ParentSchemaID 29: relation "promo_codes" (57): referenced database ID 52: referenced descriptor not found
>   ParentID  52, ParentSchemaID 29: relation "user_promo_codes" (58): constraint id was missing for constraint: FOREIGN KEY with name fk_city_ref_users
> ```

It's probably worth updating the test data to add some actual constraint IDs in there, to generate less errors.
___
*[pkg/ccl/backupccl/backup_test.go, line 2230 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0IX4DAkHxSPBw_Zj0:-Mv0IX4DAkHxSPBw_Zj1:b-asdy5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/ccl/backupccl/backup_test.go#L2230)):*
> ```Go
> 		kvDB,
> 		keys.SystemSQLCodec,
> 		version,
> ```

This pattern is annoying and noisy. For tests we should be able to just pass `nil` for the version pointer and have that default to the latest version. It's just for validation anyway, the noise isn't worth it.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 134 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv09DzrAZkMfzm3AtzX:-Mv09DzrAZkMfzm3AtzY:bj4fvx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/validate_test.go#L134)):*
> ```Go
> 			"RowLevelTTL":                   {status: iSolemnlySwearThisFieldIsValidated},
> 			"ExcludeDataFromBackup":         {status: thisFieldReferencesNoObjects},
> 			"NextConstraintID":              {status: iSolemnlySwearThisFieldIsValidated},
> ```

Do we actually validate this?
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 174 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv09I9m4wg0iy31IyH2:-Mv09I9m4wg0iy31IyH3:b-3vsmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/validate_test.go#L174)):*
> ```Go
> 			"Predicate":                   {status: iSolemnlySwearThisFieldIsValidated},
> 			"UseDeletePreservingEncoding": {status: thisFieldReferencesNoObjects},
> 			"ConstraintID":                {status: iSolemnlySwearThisFieldIsValidated},
> ```

Please add some unit test cases. One for each constraint type would be good. I'm aware it's a mildly annoying thing to do, but it's worth it.
___
*[pkg/sql/tests/repair_test.go, line 677 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv09YNfCNDsE8iDMulT:-Mv09YNfCNDsE8iDMulU:b-hig1cb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/tests/repair_test.go#L677)):*
> ```Go
>       "unique": true,
>       "version": 4,
> 			"constraintId": 1
> ```

nit: here and below: whitespace


<!-- Sent from Reviewable.io -->

fqazi committed:
schemachanger: adopt ConstraintID for comment on constraint

Previously, comments for table constraint required the use
of an OID. With the addition of the ConstraintID field for
each constraint, we will no longer be using OIDs for tracking
comments. This patch adds support for adding/deleting comments
on constraints based on the new ConstraintID field. Additionally,
the obj_description function will be updated to map OIDs to
constraint IDs.

Release note: None

postamar reviewed (changes requested):
Nicely done. This is almost ready.

> Reviewed 2 of 15 files at r22, 142 of 184 files at r34, 20 of 26 files at r35, 3 of 12 files at r36, 20 of 20 files at r37, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-:-MvPjZjU93UZg8eOqXW6:b-y8oikc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @fqazi, @rafiss, and @RichardJCai)

___
*[pkg/sql/comment_on_constraint.go, line 38 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPcWwdBuEz1CzTMbf4:-MvPcWwdBuEz1CzTMbf5:b-ysmf53) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/comment_on_constraint.go#L38)):*
> ```Go
> ) (planNode, error) {
> 	// Block comments on constraint until cluster is updated.
> 	if !p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.RemoveIncompatibleDatabasePrivileges) {
> ```

This is confusing, none of this has anything to do with database privileges. This would be a lot clearer if you defined a cluster version key named `ConstraintIDsForTableDescs` and used that instead. You don't have to make it part of the enum in `clusterversion`, you totally could define it in `tabledesc` or elsewhere as
```
// ConstraintIDsForTableDescsClusterVersion is the cluster version for which constraints in the table descriptors
// have had numerical IDs defined for them in a migration. 
const ConstraintIDsForTableDescsClusterVersion = clusterversion.RemoveIncompatibleDatabasePrivileges
```

This comment also applies elsewhere in this PR.

Furthermore, the database descriptor migration only upgrades database descriptors. You need to modify it to also upgrade table descriptors, and test it accordingly.
___
*[pkg/sql/catalog/tabledesc/structured.go, line 107 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPdSJb9FUQojbMm-FI:-MvPdSJb9FUQojbMm-FJ:b-naj7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/catalog/tabledesc/structured.go#L107)):*
> ```Go
> 	RemovedDuplicateIDsInRefs bool
> 
> 	// UpgradedConstraintID indicates that table descriptors had constraint ID
> ```

nit: s/UpgradedConstraintID/AddedConstraintIDs/ in the comment
___
*[pkg/sql/catalog/tabledesc/validate.go, line 653 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPfX1VDv2VjWlxYBwI:-MvPfX1VDv2VjWlxYBwJ:b-9qolyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/catalog/tabledesc/validate.go#L653)):*
> ```Go
> }
> 
> func (desc *wrapper) validateConstraintIDs(vea catalog.ValidationErrorAccumulator) error {
> ```

The purpose of passing `vea` here is for it to collect multiple errors, instead of returning the first one encountered. Use `vea.Report` instead of returning an error.
___
*[pkg/cli/doctor_test.go, line 84 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPePz-9sirUE7aXTjz:-MvPePz-9sirUE7aXTk-:bh69iir) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/cli/doctor_test.go#L84)):*
> ```Go
> 
> 	t.Run("deprecated doctor zipdir with verbose", func(t *testing.T) {
> 		out, err := c.RunWithCapture("debug doctor zipdir testdata/doctor/debugzip 21.1 --verbose")
> ```

Is this deliberately 21.1 or did you mean 22.1?
___
*[pkg/sql/catalog/tabledesc/helpers_test.go, line 33 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPfofV49dyemc94YRx:-MvPfofV49dyemc94YRy:b-9cudi2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/catalog/tabledesc/helpers_test.go#L33)):*
> ```Go
> 
> // Report implements catalog.ValidationErrorAccumulator
> func (*noopValidationErrorAccumulator) Report(err error) {
> ```

See my comment regarding `validateConstraintIDs`. This means you should do something with the `err` passed to this method, which in turn means you should rename `noopValidationErrorAccumulator` to `constraintValidationErrorAccumulator` or something.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 2477 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPgEJKDxQcUH-5kOPN:-MvPgEJKDxQcUH-5kOPO:byfxlln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/catalog/tabledesc/validate_test.go#L2477)):*
> ```Go
> 		desc descpb.TableDescriptor
> 	}{
> 		{`constraint id was missing for constraint: PRIMARY KEY with name primary`,
> ```

nit: use %q instead of %s for the name in the format string that produces these messages?

This way you could have a message like 
```
constraint ID was missing in PRIMARY KEY constraint "primary"
```
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 2580 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPggxiENUwXRMWFSpY:-MvPggxiENUwXRMWFSpZ:bduoeh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/catalog/tabledesc/validate_test.go#L2580)):*
> ```Go
> 			if !testutils.IsError(err, test.err) {
> 				t.Errorf(`%d: got "%v" expected "%v"`, i, err, test.err)
> 			}
> ```

FYI you might find the `require` package useful here for asserting errors (or the absence of error)


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75499)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @postamar, @rafiss, and @RichardJCai)

___
*[pkg/sql/comment_on_constraint.go, line 38 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPcWwdBuEz1CzTMbf4:-MvQ7vGZEkDrz4rVGfe0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/comment_on_constraint.go#L38)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is confusing, none of this has anything to do with database privileges. This would be a lot clearer if you defined a cluster version key named `ConstraintIDsForTableDescs` and used that instead. You don't have to make it part of the enum in `clusterversion`, you totally could define it in `tabledesc` or elsewhere as
```
// ConstraintIDsForTableDescsClusterVersion is the cluster version for which constraints in the table descriptors
// have had numerical IDs defined for them in a migration. 
const ConstraintIDsForTableDescsClusterVersion = clusterversion.RemoveIncompatibleDatabasePrivileges
```

This comment also applies elsewhere in this PR.

Furthermore, the database descriptor migration only upgrades database descriptors. You need to modify it to also upgrade table descriptors, and test it accordingly.
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 1943 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0F47_B_5pI1B-7vH0:-MvLHRup2ngajKvayhRu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/create_table.go#L1943)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

detritus?
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog.go, line 1544 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv09w7aBifCgjwJ3gxc:-MvLMOoj-IZnjDTYfetl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c2ee0704d70e65e00ee168252f1b5a359f98d59/pkg/sql/pg_catalog.go#L1544)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this comment spurious?
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog.go, line 1613 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0A4auAA4GsEaeFe9j:-MvLNleK2T_UmcWC9IaX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c2ee0704d70e65e00ee168252f1b5a359f98d59/pkg/sql/pg_catalog.go#L1613)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

It might be worth moving this if-else-block to its own helper function which takes `constraint`, the db ID, schema name and table ID, and returns an oid.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/table_elements.go, line 130 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv05PkN0TmXu6Ik0TGZ:-MvLPNWE09oi9rXkpsNI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/table_elements.go#L130)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Do we also need a `GetConstraintID()` method on the `ConstraintToUpdate` interface?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/systemschema/system.go, line 964 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MuNAt50DUnsoCRzPr60:-MvLPkYA7ZpP1KNVBmoJ:b-rhx4xg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a94ac074ca3db34743074984b48db8d4c27099f2/pkg/sql/catalog/systemschema/system.go#L964)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

should these be 1?
</blockquote></details>

No constraints inside sequences are not externally visible and hidden.



_Code quote:_
```Go
	tbl.NextConstraintID 
```
___
*[pkg/sql/catalog/tabledesc/structured.go, line 103 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv06Er92L_56vA0MYnO:-MvLPvFv07MArj3fN-XG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/structured.go#L103)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider naming this `AddedConstraintIDs` in that case?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/table_desc_builder.go, line 632 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv06lYY1_bGT5UV9Ted:-MvLQTqO6I6bke9YXBZV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/table_desc_builder.go#L632)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider defining a closure which will return the ID and increment `desc.NextConstraintID` as a side effect.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/validate.go, line 538 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv07uUoDNX6yzV5Rtsu:-MvLQwlb9LLJ_s8E7g4x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/validate.go#L538)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Consider renaming this `validateConstraintIDs`?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/validate.go, line 653 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPfX1VDv2VjWlxYBwI:-MvQ7wWB6HQOn_ZUIV0q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/catalog/tabledesc/validate.go#L653)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

The purpose of passing `vea` here is for it to collect multiple errors, instead of returning the first one encountered. Use `vea.Report` instead of returning an error.
</blockquote></details>

Done.
___
*[pkg/sql/descmetadata/metadata_updater.go, line 29 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0DQ830j-DZVeA9x-I:-MvLR4ph3jrJBEYVU70X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4f36ffed911d6db7a2426a22fc2f21ba821c396/pkg/sql/descmetadata/metadata_updater.go#L29)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

s/commentUpdater/metadataUpdater/ ?
</blockquote></details>

Done.
___
*[pkg/cli/doctor_test.go, line 84 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPePz-9sirUE7aXTjz:-MvQ7xAd8oFRHkMIW5E9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/cli/doctor_test.go#L84)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this deliberately 21.1 or did you mean 22.1?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/helpers_test.go, line 33 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPfofV49dyemc94YRx:-MvQ7xyv05iROvndtuX0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/catalog/tabledesc/helpers_test.go#L33)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

See my comment regarding `validateConstraintIDs`. This means you should do something with the `err` passed to this method, which in turn means you should rename `noopValidationErrorAccumulator` to `constraintValidationErrorAccumulator` or something.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 134 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv09DzrAZkMfzm3AtzX:-MvLRHWH9Lt2Hqfl5-bJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/validate_test.go#L134)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Do we actually validate this?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 174 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv09I9m4wg0iy31IyH2:-MvM_ecq2dL8OpKTig1Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/sql/catalog/tabledesc/validate_test.go#L174)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Please add some unit test cases. One for each constraint type would be good. I'm aware it's a mildly annoying thing to do, but it's worth it.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/tabledesc/validate_test.go, line 2580 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPggxiENUwXRMWFSpY:-MvQ8Znk4_yajgYnAsAX:bygrggp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/sql/catalog/tabledesc/validate_test.go#L2580)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

FYI you might find the `require` package useful here for asserting errors (or the absence of error)
</blockquote></details>

There isn't any nice equivalent in require, the error functions all take in an error type only and not a string. I think that's why all the other tests use this in addition to regex support.
___
*[pkg/ccl/backupccl/backup_test.go, line 2230 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0IX4DAkHxSPBw_Zj0:-MvLIJvK8imS7L22W-cg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/ccl/backupccl/backup_test.go#L2230)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This pattern is annoying and noisy. For tests we should be able to just pass `nil` for the version pointer and have that default to the latest version. It's just for validation anyway, the noise isn't worth it.
</blockquote></details>

Done.
___
*[pkg/cli/doctor.go, line 94 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0EP8S6k6ZeBEIL-dR:-MvLDmOHAblXrTIFV8Q2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/cli/doctor.go#L94)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Neat!
</blockquote></details>

Done.
___
*[pkg/cli/doctor.go, line 169 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0ESiS4OTcv9SMlLNc:-MvLDnTIAAqTqrvtxyB6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/cli/doctor.go#L169)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't the default cluster version be the last one instead?
</blockquote></details>

Done.
___
*[pkg/cli/testdata/doctor/test_examine_zipdir, line 11 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv08pusCcthxOXtR-yU:-MvLRFEICz7P0sY31qSC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4df60e58f2470d3c5117f6ea83fd06f898c64d12/pkg/cli/testdata/doctor/test_examine_zipdir#L11)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

It's probably worth updating the test data to add some actual constraint IDs in there, to generate less errors.
</blockquote></details>

Done.
___
*[pkg/migration/migrations/ensure_constraint_id.go, line 24 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0BZxm1XFMvs5wAYHr:-MvLLyTZ7bAZHgbGIA6u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69d68af2469f9904c09b6b1fb4fcf0f3146ef8b9/pkg/migration/migrations/ensure_constraint_id.go#L24)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`ensureTablesHaveConstraintIDs`?
</blockquote></details>

Done.
___
*[pkg/migration/migrations/ensure_constraint_id.go, line 63 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0BkQo5_SM1yM6ugTT:-MvLLynTBBziCHoDd0K6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69d68af2469f9904c09b6b1fb4fcf0f3146ef8b9/pkg/migration/migrations/ensure_constraint_id.go#L63)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this migration even necessary? I suspect that Richard's privileges migration will already be upgrading all table descriptors making this one entirely redundant. If that's not the case, then we should modify the existing migration to make it do what we want. Rewriting the whole descriptors table is expensive, no need to do it twice.
</blockquote></details>

Done.
___
*[pkg/sql/create_sequence.go, line 20 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0Emjw6xIesYWjJ7oy:-MvLHQ8Z3byxAjmd3zoo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/create_sequence.go#L20)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this deliberate?
</blockquote></details>

Done.
___
*[pkg/sql/create_sequence.go, line 244 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0Ev5z0-75kOPPyi1Y:-MvLHR9718jhw2RgpGWy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/create_sequence.go#L244)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This seems redundant, isn't the planner able to provide the cluster settings already?
</blockquote></details>

Done.
___
*[pkg/sql/table.go, line 298 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0FB4T9fCzcXg6-Pdh:-MvLH_Ws3jL99mlElvZP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/table.go#L298)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Can you make your whitespacing style at least somewhat consistent please? Having at least some regularity would make for a more pleasant code reading experience.
</blockquote></details>

Done.
___
*[pkg/sql/virtual_schema.go, line 697 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0GI3d2oVXMAR3KW7G:-MvLHc4x9mA8eye-itSQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/virtual_schema.go#L697)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is surprising. Is this right?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/descriptor.go, line 131 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0GiX1AfAB6rhLcvM-:-MvLHzjXEPFpOBNa89SG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/catalog/descs/descriptor.go#L131)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't the settings be a field in the `kv` struct instead?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/direct.go, line 145 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0H4MG6kvW2gXyh7m3:-MvLI-aU1i6k5nxyhNsG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/catalog/descs/direct.go#L145)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This ActiveVersion retrieval ends up being all over the place, can't we push it down closer to the validation code perhaps? 

Comment applies elsewhere also, of course.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/desctestutils/descriptor_test_utils.go, line 31 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0HKup-PBfHzmFNzFz:-MvLI0cLDlKJ3dA8KWIP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/catalog/desctestutils/descriptor_test_utils.go#L31)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Again, we almost certainly have a `var` in `clusterversion` which defines the latest version.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/desctestutils/descriptor_test_utils.go, line 70 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-Mv0HVU5DAONUszTDcdi:-MvLI2TM5AdpqQF7UCIC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd549664312d9933dd160713bc6c8b0e254fb1f8/pkg/sql/catalog/desctestutils/descriptor_test_utils.go#L70)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't these be pointers? Comment applies elsewhere also.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75499)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @postamar, @rafiss, and @RichardJCai)

___
*[pkg/cli/doctor_test.go, line 84 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvPePz-9sirUE7aXTjz:-MvQg22nBxvDCkNxTn77:b-yl4fe9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a3583f1f8749dc5e40f33295930087c6d79e767f/pkg/cli/doctor_test.go#L84)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Yes it's deliberate to void the new validation from being used on old descriptors. @postamar This was also causing the CI failures, that you observed since the new validation looks for constraint IDs. We can either leave this down level or if you want let's do a separate PR where I rework the descriptors to have constraint IDs in this test.


<!-- Sent from Reviewable.io -->

postamar reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 40 files at r38, 20 of 25 files at r39, 6 of 7 files at r40, 12 of 12 files at r41, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-:-MvQrqUjA-vX7QwKYCug:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @fqazi, @rafiss, and @RichardJCai)

___
*[pkg/cli/doctor_test.go, line 84 at r41](https://reviewable.io/reviews/cockroachdb/cockroach/75499#-MvQqscJE0in9ES18OnB:-MvQqscJE0in9ES18OnC:b657tso) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f14a74059a1413ccc0145ba5412c57044c1f0aa/pkg/cli/doctor_test.go#L84)):*
> ```Go
> 
> 	t.Run("deprecated doctor zipdir with verbose", func(t *testing.T) {
> 		out, err := c.RunWithCapture("debug doctor zipdir testdata/doctor/debugzip 22.11-52 --verbose")
> ```

22.11-52 is not a valid version number


<!-- Sent from Reviewable.io -->

fqazi commented:
bors r+
